docs/development_best_practices.md

```diff
@@ -0,0 +1,130 @@
+
+*StreamCodec* is also useful in another aspect of the application. It is used to serialize and deserialize the tuple to transfer it between operators. The default *StreamCodec* uses Kryo library for serialization. 
+
+The following guidelines are useful when considering a custom *StreamCodec*
+
+* A custom *StreamCodec* is needed if the tuples need to be distributed based on a criteria different from the hashcode of the tuple. If the correct working of an operator depends on the data from the upstream operator being distributed using a custom criteria such as being sticky on a “key” field within the tuple, then a custom *StreamCodec* should be provided by the operator developer. This codec can implement the custom criteria. The operator should also return this custom codec in the getStreamCodec method of the input port.
+* When implementing a custom *StreamCodec* for the purpose of using a different criteria to distribute the tuples, the codec can extend an existing *StreamCodec* and implement the hashcode method, so that the codec does not have to worry about the serialization and deserialization functionality. The Apex platform provides two pre-built *StreamCodec* implementations for this purpose, one is **KryoSerializableStreamCodec** that uses Kryo for serialization and another one **JavaSerializationStreamCodec** that uses JAVA serialization.
+* Different *StreamCodec* implementations can be used for different inputs in a stream with multiple inputs when different criteria of distributing the tuples is desired between the multiple inputs. 
+
+## Threads
+
+The operator lifecycle methods such as **setup**, **beginWindow**, **endWindow**, **process** in *InputPorts* all are called from a single operator lifecycle thread unbeknownst to the user. So the user does not have to worry about dealing with the issues arising from multi-threaded code. Use of separate threads in an operator is discouraged because in most cases the motivation for this is parallelism, but parallelism can already be achieved by using multiple partitions and furthermore mistakes can be made easily when writing multi-threaded code. When dealing with high volume and velocity data, the corner cases with incorrectly written multi-threaded code are encountered more easily and exposed. However, there are times when separate threads are needed, for example, when interacting with external systems the delay in retrieving or sending data can be large at times, blocking the operator and other DAG processing such as committed windows. In these cases the following guidelines must be followed strictly
+
+* Threads should be started in **activate** and closed down in deactivate. Wait till they finish in **deactivate**.
```

Comment:
Rewrite: Threads should be started in **activate**; symmetrically, they should be waited for and joined or terminated in **deactivate**.